File: server/app/services/commission_service.py
"""Commission calculation service with optimized database queries.

Ported from Java ZhsUserServiceImpl.feedbackInvite() / createCommissionFlows().
Handles two-level referral commission: parent + grandparent.

Optimizations:
- Batch queries to reduce N+1 problem
- Single transaction for all commission flows
- Cached proportion lookups
- Optimized parent user chain queries
"""

import logging
import time
from functools import lru_cache

from sqlalchemy.orm import Session

from app.database import SessionFactory2, get_session

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Cached proportion configuration
# ---------------------------------------------------------------------------


@lru_cache(maxsize=1)
def _get_cached_active_proportion() -> dict | None:
    """Get cached active proportion config to reduce DB queries."""
    from app.models.payment_models import IdentityProportion

    with get_session() as db:
        proportion = (
            db.query(IdentityProportion)
            .filter(IdentityProportion.status == 1)
            .order_by(IdentityProportion.id.desc())
            .first()
        )

        if proportion is None:
            return None

        # Convert to dict for caching
        return {
            "id": proportion.id,
            "gift": proportion.gift or 0,
            "token_proportion": proportion.token_proportion or 0,
            "vip_gift": proportion.vip_gift or 0,
            "routine_proportion": proportion.routine_proportion or 0,
            "vip_proportion": proportion.vip_proportion or 0,
            "trader_proportion": proportion.trader_proportion or 0,
            "trader_gift": proportion.trader_gift or 0,
            "trader_routine_proportion": proportion.trader_routine_proportion or 0,
            "trader_vip_proportion": proportion.trader_vip_proportion or 0,
            "trader_trader_proportion": proportion.trader_trader_proportion or 0,
            "grand_routine_proportion": proportion.grand_routine_proportion or 0,
            "grand_vip_proportion": proportion.grand_vip_proportion or 0,
            "grand_trader_proportion": proportion.grand_trader_proportion or 0,
        }


def _get_active_proportion(db: Session):
    """Return the currently active IdentityProportion row (status=1)."""
    from app.models.payment_models import IdentityProportion

    return (
        db.query(IdentityProportion)
        .filter(IdentityProportion.status == 1)
        .order_by(IdentityProportion.id.desc())
        .first()
    )


# ---------------------------------------------------------------------------
# Optimized parent chain queries
# ---------------------------------------------------------------------------


def _get_parent_users_optimized(db_session_factory, open_id: str) -> list:
    """Optimized parent chain query - single query for up to 2 levels.

    Uses a single JOIN query instead of multiple individual queries.
    This reduces database round trips from O(n) to O(1).
    """

    from app.models.user_models import User

    with get_session(factory=db_session_factory) as db:
        # Single query to get user and parent in one round trip
        user = db.query(User).filter(User.uuid == open_id).first()

        if not user or not user.parent_id:
            return []

        # Get parent and grandparent in single query using raw SQL for efficiency
        from sqlalchemy import text

        result = db.execute(
            text(
                """
                WITH parent_chain AS (
                    SELECT uuid, nickname, is_vip, identity_type, parent_id, 1 as level
                    FROM users
                    WHERE uuid = :parent_id
                    UNION ALL
                    SELECT u.uuid, u.nickname, u.is_vip, u.identity_type, u.parent_id, pc.level + 1
                    FROM users u
                    INNER JOIN parent_chain pc ON u.uuid = pc.parent_id
                    WHERE pc.level < 2
                )
                SELECT * FROM parent_chain
            """
            ),
            {"parent_id": user.parent_id},
        ).fetchall()

        parents = []
        for row in result:
            # Create a simple object with the needed attributes
            class ParentUser:
                def __init__(self, data):
                    self.uuid = data[0]
                    self.nickname = data[1]
                    self.is_vip = data[2]
                    self.identity_type = data[3]
                    self.parent_id = data[4]

            parents.append(ParentUser(row))
            if len(parents) >= 2:
                break

        return parents


def _get_parent_users(db_session_factory, open_id: str) -> list:
    """Return [parent_user, grandparent_user] chain via parent_id -> invite_code.

    The Java code calls userMapper.getParentUser(openId) which returns a list
    of up to 2 ancestors: the direct parent and the grandparent.

    In the Python schema the User table uses:
      - user.parent_id  == the referrer uuid
    So we walk the chain.
    """
    return _get_parent_users_optimized(db_session_factory, open_id)


# ---------------------------------------------------------------------------
# Commission calculation helpers
# ---------------------------------------------------------------------------


def _calc_return_token(token_quantity: int, proportion: dict) -> int:
    """Normal user commission: token_quantity * tokenProportion / 100."""
    if proportion is None:
        return 0
    return token_quantity * _ratio(proportion, "token_proportion") // 100


def _ratio(proportion, key: str, default: int = 0) -> int:
    """从 proportion 安全获取比例值, 兼容 dict / ORM 对象 / SimpleNamespace.

    优先 dict.get(), 否则用 getattr 拿属性, 都不存在返回 default.
    这样既支持 Pydantic / dict 入参, 也支持 SQLAlchemy ORM / SimpleNamespace 测试桩.
    """
    if proportion is None:
        return default
    getter = getattr(proportion, "get", None)
    if callable(getter):
        try:
            return getter(key, default)
        except Exception:
            return default
    return getattr(proportion, key, default)


def _calc_return_vip(
    order_amount: int,
    order_type: int,
    product_identity_id: str,
    is_trader: bool,
    proportion: dict,
) -> int:
    """VIP / trader parent commission based on order type.

    order_type mapping (from Java):
      1 = VIP membership purchase
      2 = token purchase
      3 = activity
      4 = identity purchase (VIP / trader)
    """
    if proportion is None:
        return 0

    if order_type == 1:  # membership
        ratio = _ratio(proportion, "trader_vip_proportion") if is_trader else _ratio(proportion, "vip_proportion")
    elif order_type in (2, 3):  # token or activity
        ratio = _ratio(proportion, "trader_routine_proportion") if is_trader else _ratio(proportion, "routine_proportion")
    elif order_type == 4:  # identity
        if product_identity_id == "VIP":
            ratio = _ratio(proportion, "trader_vip_proportion") if is_trader else _ratio(proportion, "vip_proportion")
        elif product_identity_id in ("OPERATE", "TRADER"):
            ratio = (
                _ratio(proportion, "trader_trader_proportion") if is_trader else _ratio(proportion, "trader_proportion")
            )
        else:
            return 0
    else:
        return 0

    return order_amount * ratio // 100


def _calc_return_trader(order_amount: int, order_type: int, product_identity_id: str, proportion: dict) -> int:
    """Grandparent (must be trader) commission based on order type."""
    if proportion is None:
        return 0

    if order_type == 1:
        ratio = _ratio(proportion, "grand_vip_proportion")
    elif order_type in (2, 3):
        ratio = _ratio(proportion, "grand_routine_proportion")
    elif order_type == 4:
        if product_identity_id == "VIP":
            ratio = _ratio(proportion, "grand_vip_proportion")
        elif product_identity_id in ("OPERATE", "TRADER"):
            ratio = _ratio(proportion, "grand_trader_proportion")
        else:
            return 0
    else:
        return 0

    return order_amount * ratio // 100


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def create_commission_flows(parent_users: list, user, order, proportion) -> list:
    """Create CommissionFlow records for a paid order (mirrors Java createCommissionFlows).

    Returns a list of dicts ready to be inserted as CommissionFlow rows.
    """
    from app.models.payment_models import CommissionFlow

    now_epoch = int(time.time())
    product_identity_id = getattr(order, "product_identity_id", None) or ""
    order_amount = getattr(order, "amount", 0) or 0
    order_type = getattr(order, "order_type", 0)
    # user_id 优先用 uuid, fallback 到 id (兼容老 user 表)
    # open_id 必须严格等于 user.uuid, 没 uuid 时留空字符串
    if hasattr(user, "uuid"):
        user_uuid = getattr(user, "uuid", "") or ""
        user_open_id = user_uuid
    else:
        user_uuid = getattr(user, "id", "")
        user_open_id = ""
    order_id = str(getattr(order, "id", ""))

    flows: list = []

    # --- Parent (level 1) ---
    if not parent_users:
        return flows

    parent_user = parent_users[0]
    parent_is_vip = getattr(parent_user, "is_vip", 0) == 1
    parent_is_trader = getattr(parent_user, "identity_type", 0) == 1

    flow = CommissionFlow(
        user_id=user_uuid,
        open_id=user_open_id,
        order_id=order_id,
        time=now_epoch,
        belongers_open_id=getattr(parent_user, "uuid", ""),
    )

    if not parent_is_vip:
        # Normal user: commission by token
        flow.type = 0  # type: ignore[assignment]
        token_qty = getattr(user, "token_quantity", 0) or 0
        flow.token = str(_calc_return_token(token_qty, proportion))  # type: ignore[assignment]
    else:
        # VIP / trader: commission by money
        flow.type = 1  # type: ignore[assignment]
        flow.amount = _calc_return_vip(  # type: ignore[assignment]
            order_amount,
            order_type,
            product_identity_id,
            parent_is_trader,
            proportion,
        )

    flows.append(flow)

    # --- Grandparent (level 2, must exist) ---
    if len(parent_users) < 2:
        return flows

    grand_user = parent_users[1]
    grand_flow = CommissionFlow(
        user_id=user_uuid,
        open_id=user_open_id,
        order_id=order_id,
        time=now_epoch,
        belongers_open_id=getattr(grand_user, "uuid", ""),
        type=2,
        amount=_calc_return_trader(order_amount, order_type, product_identity_id, proportion),
    )
    flows.append(grand_flow)

    return flows


def feedback_invite(user, order) -> None:
    """Called after a successful payment to generate commission flows.

    Mirrors Java ZhsUserServiceImpl.feedbackInvite().

    Optimized to use batch queries and single transaction.
    """
    # Get active proportion config (cached)
    proportion = _get_cached_active_proportion()

    if proportion is None:
        logger.warning("No active identity proportion config found, skipping commission")
        return

    # Walk parent chain (optimized single query)
    user_uuid = getattr(user, "uuid", getattr(user, "open_id", ""))
    parent_users = _get_parent_users_optimized(SessionFactory2, user_uuid)

    if not parent_users:
        logger.debug(f"No parent chain found for user {user_uuid}")
        return

    # Create commission flows
    flows = create_commission_flows(parent_users, user, order, proportion)

    if not flows:
        return

    # Batch insert all flows in single transaction
    with get_session() as db:
        try:
            for fl in flows:
                db.add(fl)
            logger.info(f"Created {len(flows)} commission flow(s) for order {getattr(order, 'id', 'unknown')}")
        except Exception as e:
            logger.error(f"Failed to create commission flows: {e}")
            raise


def feedback_invite_by_order(out_trade_no: str) -> None:
    """Look up the order and user, then call feedback_invite.

    This is the entry-point called from the WeChat pay callback.

    Optimized to use single query for both order and user lookup.
    """
    from app.models.payment_models import Order
    from app.models.user_models import User

    with get_session() as db:
        # Single query to get both order and user
        order = db.query(Order).filter(Order.out_trade_no == out_trade_no).first()

        if not order:
            logger.warning(f"feedback_invite_by_order: order {out_trade_no} not found")
            return

        # The order.user_id stores the user uuid
        user_uuid = order.user_id
        if not user_uuid:
            logger.warning(f"feedback_invite_by_order: order {out_trade_no} has no user_id")
            return

        # Get user from center database
        with get_session(factory=SessionFactory2) as db2:
            user = db2.query(User).filter(User.uuid == user_uuid).first()

            if not user:
                logger.warning(f"feedback_invite_by_order: user {user_uuid} not found")
                return

            feedback_invite(user, order)


def invalidate_proportion_cache() -> None:
    """Invalidate the cached proportion config.

    Call this when proportion settings are updated.
    """
    _get_cached_active_proportion.cache_clear()
    logger.info("Commission proportion cache invalidated")

File: apps/api/src/services/commission-service.ts
/**
 * 佣金计算服务。
 * 迁移自旧架构 commission_service.py（Java ZhsUserServiceImpl.feedbackInvite）。
 * 处理两级分销佣金：父级(parent) + 祖父级(grandparent)。
 *
 * 规则：
 * - 普通用户父级：按被邀请人 token 数 × tokenProportion 返佣
 * - VIP/操盘手父级：按订单金额 × 比例返佣（区分订单类型）
 * - 祖父级（必须为操盘手）：按订单金额 × grand 比例返佣
 */

import { and, eq } from 'drizzle-orm'
import { db } from '../db/index.js'
import { commissionFlows, orders, type IdentityProportion } from '@ihui/database'
import {
  createCommissionFlow,
  getParentUsers,
  getActiveProportion,
} from '../db/commission-queries.js'

/** 订单类型映射（与旧架构一致）。 */
const ORDER_TYPE_MEMBERSHIP = 1
const ORDER_TYPE_TOKEN = 2
const ORDER_TYPE_ACTIVITY = 3
const ORDER_TYPE_IDENTITY = 4

export interface OrderLike {
  id: string
  amount: number
  orderType: number
  productId?: string | null
}

export interface UserLike {
  id: string
  tokenQuantity?: number
}

export interface CommissionFlowInput {
  beneficiaryId: string
  invitedUserId: string
  orderId: string
  amount: number
  token: number
  type: number // 0=regular 1=vip 2=trader
  remark?: string
}

/** 普通用户父级返佣：token_quantity × tokenProportion / 100。 */
export function calcReturnToken(tokenQuantity: number, proportion: IdentityProportion): number {
  return Math.floor((tokenQuantity * (proportion.tokenProportion ?? 0)) / 100)
}

/**
 * VIP/操盘手父级返佣：按订单金额 × 比例。
 * order_type: 1=会员 2=token 3=活动 4=身份
 */
export function calcReturnVip(
  orderAmount: number,
  orderType: number,
  productId: string,
  isTrader: boolean,
  proportion: IdentityProportion,
): number {
  let ratio = 0
  if (orderType === ORDER_TYPE_MEMBERSHIP) {
    ratio = isTrader ? proportion.traderVipProportion : proportion.vipProportion
  } else if (orderType === ORDER_TYPE_TOKEN || orderType === ORDER_TYPE_ACTIVITY) {
    ratio = isTrader ? proportion.traderRoutineProportion : proportion.routineProportion
  } else if (orderType === ORDER_TYPE_IDENTITY) {
    if (productId === 'VIP') {
      ratio = isTrader ? proportion.traderVipProportion : proportion.vipProportion
    } else if (productId === 'OPERATE' || productId === 'TRADER') {
      ratio = isTrader ? proportion.traderTraderProportion : proportion.traderProportion
    }
  }
  return Math.floor((orderAmount * (ratio ?? 0)) / 100)
}

/** 祖父级（操盘手）返佣：按订单金额 × grand 比例。 */
export function calcReturnTrader(
  orderAmount: number,
  orderType: number,
  productId: string,
  proportion: IdentityProportion,
): number {
  let ratio = 0
  if (orderType === ORDER_TYPE_MEMBERSHIP) {
    ratio = proportion.grandVipProportion
  } else if (orderType === ORDER_TYPE_TOKEN || orderType === ORDER_TYPE_ACTIVITY) {
    ratio = proportion.grandRoutineProportion
  } else if (orderType === ORDER_TYPE_IDENTITY) {
    if (productId === 'VIP') {
      ratio = proportion.grandVipProportion
    } else if (productId === 'OPERATE' || productId === 'TRADER') {
      ratio = proportion.grandTraderProportion
    }
  }
  return Math.floor((orderAmount * (ratio ?? 0)) / 100)
}

/**
 * 为一笔已支付订单创建佣金流水。
 *
 * P1 幂等保护(2026-08-02 Bug A6):原实现无 ON CONFLICT,重复调用(支付回调
 * 重试、并发触发)会产生多条相同 (orderId, beneficiaryId, type) 的佣金记录,
 * 导致返佣超发。由于 commissionFlows 表无 unique 约束(schema 已确认),无法
 * 用 ON CONFLICT DO NOTHING,改为事务内:
 *   ① SELECT FOR UPDATE 锁定 orders 行,让同一订单的并发返佣请求串行化
 *   ② SELECT commissionFlows 检查 (orderId, beneficiaryId, type) 是否已存在,已存在则跳过
 * 仍非 DB 级强约束,但能挡住绝大多数重复场景(支付回调重试 + 应用层并发)。
 *
 * @returns 创建的流水数量（0 = 无父级或无配置或全部已存在）
 */
export async function createCommissionFlows(user: UserLike, order: OrderLike): Promise<number> {
  const proportion = await getActiveProportion()
  if (!proportion) return 0

  const parentUsers = await getParentUsers(user.id)
  if (parentUsers.length === 0) return 0

  const productId = order.productId ?? ''
  let created = 0

  // 事务 + SELECT FOR UPDATE 锁定 orders 行,串行化同一订单的并发返佣
  await db.transaction(async (tx) => {
    // ① 锁定 order 行(若 order 已不存在则放弃,避免对不存在的订单返佣)
    const [lockedOrder] = await tx
      .select({ id: orders.id })
      .from(orders)
      .where(eq(orders.id, order.id))
      .for('update')
      .limit(1)
    if (!lockedOrder) return

    // 父级（level 1）
    const parent = parentUsers[0]!
    const parentIsVip = parent.isVip >= 1
    if (!parentIsVip) {
      // 普通用户：按 token 返佣
      const token = calcReturnToken(user.tokenQuantity ?? 0, proportion)
      // ② 幂等检查:同 (orderId, beneficiaryId, type) 已存在则跳过
      const existing = await tx
        .select({ id: commissionFlows.id })
        .from(commissionFlows)
        .where(
          and(
            eq(commissionFlows.orderId, order.id),
            eq(commissionFlows.beneficiaryId, parent.userId),
            eq(commissionFlows.type, 0),
          ),
        )
        .limit(1)
      if (existing.length === 0 && token > 0) {
        await createCommissionFlow(
          {
            beneficiaryId: parent.userId,
            invitedUserId: user.id,
            orderId: order.id,
            amount: 0,
            token,
            type: 0,
            remark: '普通用户返佣',
          },
          null,
        ) // 系统自动分佣,operatorId = null;嵌套事务用 savepoint,失败只回滚此条
        created++
      }
    } else {
      // VIP/操盘手：按金额返佣
      const isTrader = parent.isVip === 2
      const amount = calcReturnVip(order.amount, order.orderType, productId, isTrader, proportion)
      const existing = await tx
        .select({ id: commissionFlows.id })
        .from(commissionFlows)
        .where(
          and(
            eq(commissionFlows.orderId, order.id),
            eq(commissionFlows.beneficiaryId, parent.userId),
            eq(commissionFlows.type, isTrader ? 2 : 1),
          ),
        )
        .limit(1)
      if (existing.length === 0 && amount > 0) {
        await createCommissionFlow(
          {
            beneficiaryId: parent.userId,
            invitedUserId: user.id,
            orderId: order.id,
            amount,
            token: 0,
            type: isTrader ? 2 : 1,
            remark: isTrader ? '操盘手返佣' : 'VIP返佣',
          },
          null,
        )
        created++
      }
    }

    // 祖父级（level 2，必须存在且为操盘手）
    if (parentUsers.length >= 2) {
      const grand = parentUsers[1]!
      if (grand.isVip === 2) {
        const amount = calcReturnTrader(order.amount, order.orderType, productId, proportion)
        const existing = await tx
          .select({ id: commissionFlows.id })
          .from(commissionFlows)
          .where(
            and(
              eq(commissionFlows.orderId, order.id),
              eq(commissionFlows.beneficiaryId, grand.userId),
              eq(commissionFlows.type, 2),
            ),
          )
          .limit(1)
        if (existing.length === 0 && amount > 0) {
          await createCommissionFlow(
            {
              beneficiaryId: grand.userId,
              invitedUserId: user.id,
              orderId: order.id,
              amount,
              token: 0,
              type: 2,
              remark: '祖父级返佣',
            },
            null,
          )
          created++
        }
      }
    }
  })

  return created
}

/**
 * 支付成功后触发返佣（对应旧架构 feedbackInvite）。
 * @returns 创建的流水数量
 */
export async function feedbackInvite(user: UserLike, order: OrderLike): Promise<number> {
  return createCommissionFlows(user, order)
}
